# Grundlagen

[Video](https://youtu.be/nvyklCSuilw) - [Video2]()



## GL 1 · Variablen und Zuweisungen

### Inhalte

Zunächst betrachten wir zwei Arten von Zahlen:

| Datentyp | In Python | Beispiele |
|:--|:--|:--|
| **Ganzzahl** | `int` | `8`, `-3`, `0` |
| **Dezimalzahl** (Gleitkommazahl, *floating point number*) | `float` | `3.2`, `-0.5`, `2.0` |

**Dezimalzahlen werden mit einem Punkt geschrieben**, nicht mit einem Komma.


Mit einer **Zuweisung** ordnen wir einer Variablen einen Wert zu. Das Gleichheitszeichen ist dabei keine mathematische Aussage: Rechts wird gerechnet, links wird gespeichert.

In [1]:
k = 8          # int
x = 3.2        # float

Mit der Funktion `print` geben wir Werte aus. `print` kann – mit Komma getrennt – beliebig viele Werte ausgeben. In der Ausgabe stehen sie durch ein Leerzeichen getrennt.

In [2]:
print(k)           # 8
print(x)           # 3.2
print(k, x, 10)    # 8 3.2 10

8
3.2
8 3.2 10


Durch eine **erneute Zuweisung** verändern wir den Wert einer Variablen. Der alte Wert ist danach verloren.

In [3]:
x = 5
x = x + 1      # liest 5, rechnet 6, speichert 6
print(x)       # 6

6


Für besonders häufige Zuweisungen gibt es **Kurzformen**:

In [4]:
x += 1    # entspricht  x = x + 1
x -= 1    # entspricht  x = x - 1
x *= 2    # entspricht  x = x * 2   … usw.

### Aufgabe 1

Welche Werte haben `a`, `b` und `c` am Ende? Was gibt das Programm aus? Löse ohne Computer und lege dazu eine Tabelle an, die nach jeder Zeile die Werte festhält.

In [ ]:
a = 4
b = a * 2
a = a + 3
c = a + b
b += c
print(a, b, c)

#### Lösung

| Zeile | Anweisung | a | b | c |
|:--|:--|:--|:--|:--|
| 1 | `a = 4` | 4 | – | – |
| 2 | `b = a * 2` | 4 | 8 | – |
| 3 | `a = a + 3` | 7 | 8 | – |
| 4 | `c = a + b` | 7 | 8 | 15 |
| 5 | `b += c` | 7 | 23 | 15 |

Ausgabe: `7 23 15`

**Wichtig ist Zeile 3:** `b` bleibt 8, obwohl sich `a` ändert. Die Zuweisung `b = a * 2` rechnet *einmal* mit dem damaligen Wert von `a` und speichert das Ergebnis. Sie stellt keine dauerhafte Beziehung zwischen `a` und `b` her.

### Aufgabe 2

Die Werte zweier Variablen sollen getauscht werden. Danach soll `a` den Wert 9 und `b` den Wert 3 haben.

In [5]:
a = 3
b = 9
a = b
b = a
print(a, b)

9 9


1. Was gibt das Programm aus? Warum gelingt der Tausch nicht?
2. Korrigiere das Programm mit Hilfe einer dritten Variablen.

#### Lösung

**1.** Ausgabe: `9 9`. Nach `a = b` ist der alte Wert 3 von `a` überschrieben und damit verloren. `b = a` kopiert dann nur die 9 zurück.

**2.** Man merkt sich den alten Wert vorher in einer Hilfsvariablen:

In [6]:
a = 3
b = 9
hilf = a       # hilf = 3
a = b          # a = 9
b = hilf       # b = 3
print(a, b)    # 9 3

9 3


*Zum Weiterlesen:* Python erlaubt auch die Kurzform `a, b = b, a`. Dabei wird die rechte Seite vollständig ausgewertet, bevor zugewiesen wird.

## GL 2 · Kommentare und Variablennamen

### Inhalte

**Kommentare** werden nicht ausgeführt. Sie dienen zur Dokumentation des Programms.

In [ ]:
'''
Dies ist ein
mehrzeiliger Kommentar
'''
x = 1

# dies ist ein einzeiliger Kommentar
y = 2
z = 3   # auch dies ist ein Kommentar

**Regeln für Variablennamen:**

- Variablennamen sind **case-sensitiv**: `anzahl` und `Anzahl` sind zwei verschiedene Variablen.
- Das erste Zeichen ist ein **Buchstabe oder ein Unterstrich**, keine Ziffer.
- Wir beschränken uns freiwillig auf den **Unterstrich, die Buchstaben des englischen Alphabets und die Ziffern**. Leerzeichen und Zeichen wie `-` sind nicht erlaubt.
- Schlüsselwörter der Sprache wie `if`, `for` oder `while` sind als Namen verboten.

Wenn es der Lesbarkeit nicht schadet, erlauben wir uns kurze Namen wie `x` oder `k`. Ansonsten nehmen wir **sprechende Namen**. **Variablennamen beginnen in der Regel mit einem kleinen Buchstaben.** Zusammengesetzte Namen schreiben wir entweder in *camelCase* oder mit Unterstrich verbunden (*snake_case*), aber nicht gemischt:

In [ ]:
zaehl = 1
radius = 2.2
anzahlTreffer = 5            # camelCase
entfernungZurWand = 3.2      # camelCase
anzahl_treffer = 5           # snake_case
entfernung_zur_wand = 3.2    # snake_case

### Aufgabe 1

Welche Variablennamen entsprechen unseren Konventionen? Begründe bei den übrigen, was nicht stimmt, und gib einen besseren Namen an.

```
A)  anzahl                F)  x2
B)  2teZahl               G)  entfernung_zur_wand
C)  anzahlTreffer         H)  entfernung_zurWand
D)  mein-name             I)  preis in euro
E)  MeinName              J)  höhe
```

#### Lösung

|   | Name | Konform? | Begründung | Besser |
|:--|:--|:--|:--|:--|
| A | `anzahl` | ja | sprechend, beginnt klein |  |
| B | `2teZahl` | **nein** | beginnt mit einer Ziffer, ist in Python sogar *ungültig* | `zweiteZahl` |
| C | `anzahlTreffer` | ja | camelCase, beginnt klein |  |
| D | `mein-name` | **nein** | *ungültig*: Python liest das als `mein` minus `name` | `meinName` |
| E | `MeinName` | **nein** | gültig, beginnt aber mit einem Großbuchstaben | `meinName` |
| F | `x2` | ja | kurzer Name, erlaubt, wenn die Lesbarkeit nicht leidet |  |
| G | `entfernung_zur_wand` | ja | snake_case, beginnt klein |  |
| H | `entfernung_zurWand` | **nein** | gültig, mischt aber snake_case und camelCase | `entfernung_zur_wand` oder `entfernungZurWand` |
| I | `preis in euro` | **nein** | *ungültig*: Leerzeichen sind nicht erlaubt | `preisInEuro` oder `preis_in_euro` |
| J | `höhe` | **nein** | Python erlaubt Umlaute, wir beschränken uns aber auf das englische Alphabet | `hoehe` |

*Zum Unterschied:* Bei B, D und I würde Python einen *SyntaxError* melden. E, H und J funktionieren, verstoßen aber gegen unsere Regeln für gut lesbare Programme.

### Aufgabe 2

1. Gib für die folgenden Größen je einen sprechenden Namen in *camelCase* und in *snake_case* an: die Anzahl der Schüler einer Klasse, die Höhe eines Dreiecks, die Durchschnittsnote einer Klassenarbeit.
2. Was gibt das folgende Programm aus? Erkläre.

In [9]:
wert = 3
Wert = 5
print(wert)

3


#### Lösung

**1.**

| Größe | camelCase | snake_case |
|:--|:--|:--|
| Anzahl der Schüler | `anzahlSchueler` | `anzahl_schueler` |
| Höhe des Dreiecks | `hoeheDreieck` | `hoehe_dreieck` |
| Durchschnittsnote der Klassenarbeit | `durchschnittKlassenarbeit` | `durchschnitt_klassenarbeit` |

Die Umlaute ersetzen wir durch *ae*, *oe* und *ue*, weil wir uns auf das englische Alphabet beschränken. In einem Programm sollte man sich für *eine* der beiden Schreibweisen entscheiden und sie durchgehend verwenden.

**2.** Ausgabe: `3`. Weil Variablennamen case-sensitiv sind, sind `wert` und `Wert` zwei verschiedene Variablen. Die Zuweisung in Zeile 2 ändert `wert` nicht. Gerade deshalb sind solche fast gleichen Namen eine Fehlerquelle und sollten vermieden werden.

## GL 3 · Arithmetische Operatoren

### Inhalte

Neben den Grundrechenarten kennt Python die Exponentation, den Modulo-Operator und die ganzzahlige Division.

| Operator | Bedeutung | Beispiel | Ergebnis |
|:--|:--|:--|:--|
| `+` | Addition | `11 + 3` | 14 |
| `-` | Subtraktion | `11 - 3` | 8 |
| `*` | Multiplikation | `11 * 3` | 33 |
| `/` | Division | `11 / 3` | 3.6666666666666665 |
| `**` | Exponentation (11 hoch 3) | `11 ** 3` | 1331 |
| `%` | Modulo: Rest bei der Division | `11 % 3` | 2 |
| `//` | ganzzahlige Division | `11 // 3` | 3 |

**`//` und `%` gehören zusammen:** 11 geteilt durch 3 ergibt 3 Rest 2, denn 11 = 3 · 3 + 2.

**Die Division `/` liefert immer einen `float`**, auch wenn sie aufgeht: `6 / 2` ergibt `3.0`.

**Auswertungsreihenfolge:** Exponentation vor Punktrechnung vor Strichrechnung. Im Zweifelsfall setzen wir Klammern.

In [11]:
print(2 + 3 * 4)       # 14   Punkt vor Strich
print((2 + 3) * 4)     # 20   Klammern zuerst
print(2 * 3 ** 2)      # 18   Exponentation vor Punktrechnung

14
20
18


### Aufgabe 1

Berechne ohne Computer. Gib bei jedem Ergebnis an, ob es ein `int` oder ein `float` ist.

```
A)  17 // 5         F)  2 + 3 ** 2
B)  17 % 5          G)  20 / 4
C)  2 ** 4          H)  7 % 2
D)  10 - 2 * 3      I)  8 % 2
E)  (10 - 2) * 3
```

#### Lösung

|   | Ausdruck | Ergebnis | Typ | Erläuterung |
|:--|:--|:--|:--|:--|
| A | `17 // 5` | 3 | int | 17 = 3 · 5 + 2 |
| B | `17 % 5` | 2 | int | der Rest dazu |
| C | `2 ** 4` | 16 | int | 2 · 2 · 2 · 2 |
| D | `10 - 2 * 3` | 4 | int | Punkt vor Strich: 10 − 6 |
| E | `(10 - 2) * 3` | 24 | int | Klammer zuerst: 8 · 3 |
| F | `2 + 3 ** 2` | 11 | int | Exponentation zuerst: 2 + 9 |
| G | `20 / 4` | 5.0 | **float** | `/` liefert immer float |
| H | `7 % 2` | 1 | int | 7 ist ungerade |
| I | `8 % 2` | 0 | int | 8 ist gerade |

**H und I zeigen einen wichtigen Trick:** Eine Zahl ist genau dann gerade, wenn `zahl % 2` den Wert 0 ergibt.

In [14]:
# Zum Nachprüfen
print(17 // 5, 17 % 5, 2 ** 4, 10 - 2 * 3, (10 - 2) * 3)
print(2 + 3 ** 2, 20 / 4, 7 % 2, 8 % 2)

3 2 16 4 24
11 5.0 1 0


### Aufgabe 2

1. Eine Zeitangabe von `sekunden = 4000` soll in Stunden, Minuten und Sekunden umgerechnet werden. Schreibe ein Programm, das `//` und `%` verwendet.
2. Wie erhält man von `zahl = 4711` die letzte Ziffer? Wie erhält man die Zahl ohne die letzte Ziffer?

#### Lösung

**1.** Eine Stunde hat 3600 Sekunden, eine Minute 60.

In [12]:
sekunden = 4000
stunden = sekunden // 3600     # 1
rest = sekunden % 3600         # 400 Sekunden bleiben übrig
minuten = rest // 60           # 6
sek = rest % 60                # 40
print(stunden, minuten, sek)   # 1 6 40

1 6 40


4000 Sekunden sind also 1 Stunde, 6 Minuten und 40 Sekunden. *Probe:* 3600 + 6 · 60 + 40 = 4000 ✓

**2.** Dividiert man durch 10, ist der Rest die letzte Ziffer:

In [13]:
zahl = 4711
print(zahl % 10)     # 1     letzte Ziffer
print(zahl // 10)    # 471   Zahl ohne die letzte Ziffer

1
471


## GL 4 · Strings: Länge und Konkatenation

### Inhalte

**Strings** sind Zeichenketten. Sie werden in einfache `'…'` oder doppelte `"…"` Hochkommata eingeschlossen. Beide Schreibweisen sind gleichwertig.

In [16]:
a = 'Hallo Welt'
b = "Guten Morgen"
print(a)          # Hallo Welt
print(b)          # Guten Morgen

Hallo Welt
Guten Morgen


Die **Länge** eines Strings ist die Anzahl seiner Zeichen. Wir erhalten sie mit `len()`. Leerzeichen zählen mit.

In [17]:
s = 'Habicht'
print(len(s))     # 7

7


Mit dem Zeichen `+` können wir Strings aneinanderhängen (**konkatenieren**). Leerzeichen werden dabei nicht automatisch eingefügt, wir müssen sie selbst ergänzen.

In [18]:
s1 = 'Hallo'
s2 = 'du'
s3 = 'Welt'
s4 = s1 + ' ' + s2 + ' ' + s3
print(s4)         # Hallo du Welt

Hallo du Welt


**Vorsicht:** `'3' + '4'` ergibt den String `'34'`, denn Strings werden verkettet, nicht addiert. Einen String und eine Zahl kann man nicht mit `+` verbinden: `'Alter: ' + 15` führt zu einem *TypeError*.

### Aufgabe 1

Was gibt das Programm aus?

In [ ]:
vorname = 'Ada'
nachname = 'Lovelace'
print(vorname + nachname)
print(vorname + ' ' + nachname)
print(len(vorname + nachname))
print(len('Hallo Welt'))
print('3' + '4')

#### Lösung

```
AdaLovelace
Ada Lovelace
11
10
34
```

- Ohne `' '` kleben die beiden Namen direkt aneinander.
- `'AdaLovelace'` hat 3 + 8 = 11 Zeichen.
- `'Hallo Welt'` hat 10 Zeichen, denn das Leerzeichen zählt mit.
- `'3' + '4'` verkettet zwei Strings zu `'34'`. Das Ergebnis ist keine Zahl.

### Aufgabe 2

Für `wort = 'Informatik'` soll das Programm den Satz `Informatik hat 10 Zeichen.` ausgeben. Die Zahl soll dabei berechnet werden.

1. Warum scheitert `print(wort + ' hat ' + len(wort) + ' Zeichen.')`?
2. Gib eine Lösung an, die `print` mit mehreren Werten verwendet.

#### Lösung

**1.** `len(wort)` liefert die Zahl 10, also einen `int`. Ein String und eine Zahl lassen sich nicht mit `+` verbinden, Python meldet einen *TypeError*.

**2.** `print` mit Komma gibt beliebige Werte aus und setzt Leerzeichen dazwischen:

In [19]:
wort = 'Informatik'
print(wort, 'hat', len(wort), 'Zeichen.')
# Informatik hat 10 Zeichen.

Informatik hat 10 Zeichen.


*Alternativ* wandelt man die Zahl mit `str()` in einen String um: `wort + ' hat ' + str(len(wort)) + ' Zeichen.'`. Am elegantesten ist ein f-String (siehe GL 8).

## GL 5 · Strings: Indexing und Slicing

### Inhalte

Jedes Zeichen eines Strings hat **zwei Indizes**: von vorne gezählt positiv, beginnend bei `0`, und von hinten gezählt negativ, beginnend bei `-1`.

<img src="./habicht.png" width="300" alt="Das Wort Habicht mit den positiven Indizes 0 bis 6 und den negativen Indizes -7 bis -1">

```
  H   a   b   i   c   h   t
  0   1   2   3   4   5   6
 -7  -6  -5  -4  -3  -2  -1
```

Mit **Indexing** greifen wir auf ein einzelnes Zeichen zu:

In [20]:
s = 'Habicht'
print(s[0])           # H   erstes Zeichen
print(s[1])           # a   zweites Zeichen
print(s[len(s)-1])    # t   letztes Zeichen
print(s[-1])          # t   letztes Zeichen, kürzer
print(s[-2])          # h   vorletztes Zeichen

H
a
t
t
h


Der größte positive Index ist `len(s) - 1`. Ein Zugriff wie `s[7]` führt zu einem *IndexError*.

Mit **Slicing** erzeugen wir Teilstrings: `s[von:bis]`. Fehlt die von-Zahl, beginnt der Teilstring am Anfang. Fehlt die bis-Zahl, reicht er bis zum Ende.

In [21]:
s = 'Habicht'
print(s[:3])          # Hab     die ersten 3 Zeichen
print(s[-3:])         # cht     die letzten 3 Zeichen
print(s[2:])          # bicht   alles außer den ersten beiden Zeichen
print(s[:-2])         # Habic   alles außer den letzten beiden Zeichen

Hab
cht
bicht
Habic


**Der Index der bis-Zahl wird nicht mitgenommen.** Die Differenz `bis - von` ist deshalb genau die Länge des Ergebnisses.

In [22]:
s = 'Habicht'
print(s[1:5])         # abic    Index 1 bis 4, Länge 5 - 1 = 4

abic


### Aufgabe 1

Gegeben ist `s = 'Programmieren'`. Was ergeben die folgenden Ausdrücke?

```
A)  s[0]          F)  s[7:]
B)  s[3]          G)  s[3:6]
C)  s[-1]         H)  s[-4:]
D)  s[-3]         I)  len(s)
E)  s[:7]         J)  s[13]
```

#### Lösung

Am besten schreibt man sich die Indizes zuerst auf:

```
  P   r   o   g   r   a   m   m   i   e   r   e   n
  0   1   2   3   4   5   6   7   8   9  10  11  12
-13 -12 -11 -10  -9  -8  -7  -6  -5  -4  -3  -2  -1
```

|   | Ausdruck | Ergebnis | Erläuterung |
|:--|:--|:--|:--|
| A | `s[0]` | `'P'` | erstes Zeichen |
| B | `s[3]` | `'g'` | viertes Zeichen |
| C | `s[-1]` | `'n'` | letztes Zeichen |
| D | `s[-3]` | `'r'` | drittletztes Zeichen |
| E | `s[:7]` | `'Program'` | Index 0 bis 6 |
| F | `s[7:]` | `'mieren'` | ab Index 7 bis zum Ende |
| G | `s[3:6]` | `'gra'` | Index 3, 4, 5 – Länge 6 − 3 = 3 |
| H | `s[-4:]` | `'eren'` | die letzten 4 Zeichen |
| I | `len(s)` | 13 | 13 Zeichen |
| J | `s[13]` | *IndexError* | der größte Index ist 12 |

*Beobachtung zu E und F:* `s[:7] + s[7:]` ergibt wieder den ganzen String. An der Stelle 7 wird der String sauber in zwei Teile zerlegt, kein Zeichen fehlt und keines kommt doppelt vor.

In [ ]:
# Zum Nachprüfen
s = 'Programmieren'
print(s[0], s[3], s[-1], s[-3])
print(s[:7], s[7:], s[3:6], s[-4:])
print(len(s))
print(s[13])        # IndexError

### Aufgabe 2

Gegeben ist `w = 'Fussballweltmeisterschaft'`.

1. Gib Slicing-Ausdrücke an, die die Teilwörter `'ball'`, `'welt'`, `'meister'` und `'schaft'` liefern.
2. Gib einen Ausdruck an, der für *jeden* String alles außer dem ersten und dem letzten Zeichen liefert.

#### Lösung

**1.** Der String hat 25 Zeichen:

```
F  u  s  s  b  a  l  l  w  e  l  t  m  e  i  s  t  e  r  s  c  h  a  f  t
0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
```

In [ ]:
w = 'Fussballweltmeisterschaft'
print(w[4:8])       # ball
print(w[8:12])      # welt
print(w[12:19])     # meister
print(w[-6:])       # schaft    oder  w[19:]

Bei jedem Teilwort ist die bis-Zahl die von-Zahl des nächsten Teilworts. Das liegt daran, dass die bis-Zahl nicht mitgenommen wird.

**2.** `s[1:-1]` beginnt beim zweiten Zeichen und hört vor dem letzten Zeichen auf. Für `'Habicht'` ergibt das `'abich'`.

## GL 6 · Strings: Methoden und Vergleiche

### Inhalte

Mit den String-Methoden `lower()` und `upper()` verwandeln wir alle Buchstaben in Klein- bzw. Großbuchstaben. Sie liefern einen **neuen String**, der ursprüngliche String bleibt unverändert.

In [23]:
s = 'Habicht'
print(s.lower())    # habicht
print(s.upper())    # HABICHT
print(s)            # Habicht   unverändert

habicht
HABICHT
Habicht


Mit dem Zeichen `<` prüfen wir die **alphabetische Ordnung**. Verglichen wird Zeichen für Zeichen, bis sich zwei Zeichen unterscheiden. Dabei gilt: **Großbuchstaben < Kleinbuchstaben**.

In [24]:
print('aab' < 'abb')    # True    das zweite Zeichen entscheidet
print('a' < 'A')        # False   Großbuchstaben kommen zuerst

True
False


Mit dem Schlüsselwort `in` prüfen wir, ob ein Zeichen oder ein Teilstring in einem anderen String enthalten ist.

In [25]:
print('a' in 'Habicht')    # True
print('e' in 'Habicht')    # False

True
False


Mit der String-Methode `find()` finden wir den **Index des ersten Auftretens** eines Teilstrings. Das Ergebnis `-1` heißt: Es wurde nichts gefunden.

In [26]:
s = 'Habicht'
print(s.find('ich'))    # 3
print(s.find('b'))      # 2
print(s.find('e'))      # -1

3
2
-1


Spätere Auftreten finden wir, wenn wir als zweiten Wert mitgeben, **ab welchem Index** gesucht werden soll:

In [27]:
s = 'aaabbaaaabbaaaa'
print(s.find('bb'))       # 3
print(s.find('bb', 4))    # 9    suche 'bb' ab Index 4

3
9


### Aufgabe 1

Was gibt das Programm aus?

In [ ]:
s = 'Banane'
print(s.upper())
print(s)
print('nan' in s)
print('N' in s)
print(s.find('an'))
print(s.find('an', 2))
print(s.find('x'))
print('Banane' < 'Birne')
print('banane' < 'Birne')

#### Lösung

| Anweisung | Ausgabe | Erläuterung |
|:--|:--|:--|
| `s.upper()` | `BANANE` | neuer String in Großbuchstaben |
| `s` | `Banane` | `upper()` hat `s` nicht verändert |
| `'nan' in s` | `True` | B-a-**nan**-e |
| `'N' in s` | `False` | Groß- und Kleinbuchstaben sind verschiedene Zeichen |
| `s.find('an')` | `1` | erstes Auftreten ab Index 1 |
| `s.find('an', 2)` | `3` | ab Index 2 gesucht: das zweite `'an'` |
| `s.find('x')` | `-1` | nicht gefunden |
| `'Banane' < 'Birne'` | `True` | erstes Zeichen gleich, dann `'a' < 'i'` |
| `'banane' < 'Birne'` | `False` | `'B'` ist als Großbuchstabe kleiner als `'b'` |

### Aufgabe 2

1. Für `satz = 'Der Habicht jagt eine Maus'` ergibt `'habicht' in satz` den Wert `False`. Wie kann man prüfen, ob das Wort enthalten ist, ohne auf Groß- und Kleinschreibung zu achten?
2. `'apfel' < 'Birne'` ergibt `False`, obwohl *Apfel* im Wörterbuch vor *Birne* steht. Wie vergleicht man richtig?
3. In `t = 'Hallo Welt, hallo Mond'` soll der Index des *zweiten* Auftretens von `'allo'` bestimmt werden. Gib die Anweisungen an.

#### Lösung

**1.** Man wandelt den Satz vorher in Kleinbuchstaben um:

In [28]:
satz = 'Der Habicht jagt eine Maus'
print('habicht' in satz)            # False
print('habicht' in satz.lower())    # True

False
True


`satz.lower()` ergibt `'der habicht jagt eine maus'`, darin ist `'habicht'` enthalten.

**2.** Auch hier hilft es, beide Wörter auf dieselbe Schreibweise zu bringen:

In [29]:
print('apfel'.lower() < 'Birne'.lower())    # True

True


**3.** Man sucht das erste Auftreten und sucht dann ab dem Index dahinter weiter:

In [30]:
t = 'Hallo Welt, hallo Mond'
erster = t.find('allo')              # 1
zweiter = t.find('allo', erster + 1) # 13
print(zweiter)

13


Würde man ab `erster` statt ab `erster + 1` suchen, fände `find()` wieder das erste Auftreten bei Index 1.

## GL 7 · Eingabe mit `input()`

### Inhalte

Mit der Funktion `input()` lesen wir Eingaben des Benutzers von der Konsole ein. Der Text in den Klammern wird als Aufforderung angezeigt. Das Programm wartet, bis der Benutzer die Eingabe mit Enter abschließt.

In [31]:
s = input('Wie ist dein Vorname? ')
print('Hallo', s)

Wie ist dein Vorname?  Malte


Hallo Malte


```
Wie ist dein Vorname? Malte
Hallo Malte
```

**`input()` liefert immer einen String**, auch wenn Ziffern eingegeben werden. Wollen wir eine Eingabe als Zahl behandeln, wandeln wir den String mit `int()` oder `float()` in eine Zahl um.

In [ ]:
x = int(input('Bitte eine ganze Zahl eingeben: '))
print('Das Doppelte von', x, 'ist', 2 * x)

```
Bitte eine ganze Zahl eingeben: 12
Das Doppelte von 12 ist 24
```

Ist die Eingabe keine gültige Zahl, etwa `abc` oder bei `int()` eine Dezimalzahl wie `3.5`, bricht das Programm mit einem *ValueError* ab. Auch `float()` erwartet einen Punkt und kein Komma.

### Aufgabe 1

Das folgende Programm soll das Alter in fünf Jahren ausgeben:

In [ ]:
alter = input('Wie alt bist du? ')
print('In 5 Jahren bist du', alter + 5)

1. Das Programm bricht mit einem Fehler ab. Erkläre, warum.
2. Korrigiere das Programm.
3. Was passiert im korrigierten Programm, wenn jemand `15.5` eingibt?

#### Lösung

**1.** `alter` ist ein String, zum Beispiel `'15'`. Einen String und eine Zahl kann man nicht mit `+` verbinden, Python meldet einen *TypeError*.

**2.** Die Eingabe wird gleich beim Einlesen in eine Zahl umgewandelt:

In [ ]:
alter = int(input('Wie alt bist du? '))
print('In 5 Jahren bist du', alter + 5)

**3.** `int('15.5')` ist nicht möglich, das Programm bricht mit einem *ValueError* ab. Sollen auch Dezimalzahlen erlaubt sein, verwendet man `float()` statt `int()`.

### Aufgabe 2

Schreibe ein Programm, das die Länge und die Breite eines Rechtecks einliest und den Flächeninhalt und den Umfang ausgibt. Auch Dezimalzahlen sollen möglich sein.

#### Lösung

In [38]:
laenge = float(input('Länge: '))
breite = float(input('Breite: '))
flaeche = laenge * breite
umfang = 2 * (laenge + breite)
print('Flächeninhalt:', flaeche)
print('Umfang:', umfang)

Länge:  4.5
Breite:  2


Flächeninhalt: 9.0
Umfang: 13.0


```
Länge: 4.5
Breite: 2
Flächeninhalt: 9.0
Umfang: 13.0
```

`float()` wandelt auch ganze Zahlen um: Aus der Eingabe `2` wird `2.0`. Deshalb sind beide Ergebnisse Dezimalzahlen.

Beim Umfang sind die **Klammern nötig**. Ohne sie würde `2 * laenge + breite` nur die Länge verdoppeln.

## GL 8 · Ausgabe mit f-Strings

### Inhalte

Für die Ausgabe von Text und Variablen nutzen wir einen **f-String**. Er beginnt mit einem `f` vor dem Hochkomma. Variablen oder ganze Ausdrücke schreiben wir in **geschweiften Klammern** direkt in den String. Eine Umwandlung mit `str()` ist nicht nötig.

In [39]:
x = int(input('Bitte eine ganze Zahl eingeben: '))
print(f'Das Doppelte von {x} ist {2*x}.')

Bitte eine ganze Zahl eingeben:  12


Das Doppelte von 12 ist 24.


```
Bitte eine ganze Zahl eingeben: 12
Das Doppelte von 12 ist 24.
```

Bei Bedarf fügen wir nach einem Doppelpunkt eine **Formatierung** hinzu. Mit `:.2f` wird eine Dezimalzahl auf zwei Nachkommastellen gerundet:

In [40]:
pi = 3.1415926535
print(f'Pi auf zwei Nachkommastellen: {pi:.2f}')
# Pi auf zwei Nachkommastellen: 3.14

Pi auf zwei Nachkommastellen: 3.14


Die Zahl vor dem `f` gibt die Anzahl der Nachkommastellen an: `{pi:.4f}` ergibt `3.1416`. Der Wert der Variablen selbst bleibt dabei unverändert, gerundet wird nur die Ausgabe.

### Aufgabe 1

1. Was gibt `print(f'{2/3:.2f}')` aus? Wird gerundet oder abgeschnitten?
2. Gegeben sind `name = 'Lea'`, `punkte = 17` und `maxPunkte = 20`. Gib mit einem f-String den Satz `Lea hat 17 von 20 Punkten erreicht (85.0 %).` aus. Der Prozentwert soll berechnet und mit einer Nachkommastelle ausgegeben werden.

#### Lösung

**1.** Ausgabe: `0.67`. `2/3` ist 0.6666…, die Formatierung **rundet** auf zwei Stellen. Abgeschnitten wäre es 0.66.

**2.**

In [41]:
name = 'Lea'
punkte = 17
maxPunkte = 20
print(f'{name} hat {punkte} von {maxPunkte} Punkten erreicht ({punkte/maxPunkte*100:.1f} %).')

Lea hat 17 von 20 Punkten erreicht (85.0 %).


In den geschweiften Klammern darf ein ganzer Ausdruck stehen. Hier wird der Prozentwert 17 / 20 · 100 = 85.0 direkt im f-String berechnet und mit `:.1f` formatiert.

### Aufgabe 2

Schreibe einen Preisrechner: Das Programm liest einen Nettopreis ein und gibt die Mehrwertsteuer (19 %) und den Bruttopreis aus, jeweils mit zwei Nachkommastellen und dem Zusatz `Euro`.

#### Lösung

In [42]:
netto = float(input('Nettopreis in Euro: '))
mwst = netto * 0.19
brutto = netto + mwst
print(f'Mehrwertsteuer: {mwst:.2f} Euro')
print(f'Bruttopreis:    {brutto:.2f} Euro')

Nettopreis in Euro:  49.99


Mehrwertsteuer: 9.50 Euro
Bruttopreis:    59.49 Euro


```
Nettopreis in Euro: 49.99
Mehrwertsteuer: 9.50 Euro
Bruttopreis:    59.49 Euro
```

Ohne Formatierung erschiene `9.4981` bzw. `59.4881`. Bei Geldbeträgen sind aber zwei Nachkommastellen üblich. Mit `:.2f` erscheint außerdem `9.50` statt `9.5`, die Null am Ende bleibt also stehen.